### ISSUE TITLE
Consider not emitting CS1998 for interface implementations / method overrides

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#77001]
CS1998 warns about `async` methods that lack any `await`s. There's a balance to be struck here, as there are valid scenarios for this to happen, especially when you're forced to have a Task-returning method due to external requirements, e.g. you're overriding a Task-returning method or you're implementing an interface method. I end up seeing a non-trivial amount of code that gets this warning in such cases, and then rather than suppressing it, they add in an unnecessary await, e.g. `await Task.FromResult(result)`, which doesn't make the method any "safer" but does make it more expensive.

I suggest we consider not emitting CS1998 when the `async` method in question is an override or is part of implementing an interface.

cc: @jcouv

**Comment by CyrusNajmabadi** [comment]
Perhaps a different warning id that can be suppressed independently?

**Comment by stephentoub** [comment]
> Perhaps a different warning id that can be suppressed independently?

If we could make it a suggestion/ info by default, sure. My concern is the default warning is leading some folks to the wrong default outcome.

**Comment by jcouv** [comment]
Why wouldn't the user remove the `async` in such case and just implement as a normal Task-returning method (ie. return `Task.FromResult(...)` without `await`) to satisfy the interface/override?
Maybe an example would be useful.
We could adjust the diagnostic message to make it clear that is a valid option.

**Comment by stephentoub** [comment]
> Why wouldn't the user remove the `async` in such case and just implement as a normal Task-returning method (ie. return `Task.FromResult(...)` without `await`) to satisfy the interface/override? Maybe an example would be useful. We could adjust the diagnostic message to make it clear that is a valid option.

Just using Task.FromResult is often "incorrect", e.g. if I have:
```C#
public override Task<Stream> GetStreamAsync()
{
    return Task.FromResult<Stream>(new FileStream(...));
}
```
that fails to catch any exceptions from the FileStream ctor and store them into the returned task, so it really need to be more like:
```C#
public override Task<Stream> GetStreamAsync()
{
    try
    {
        return Task.FromResult<Stream>(new FileStream(...));
    }
    catch (Exception e)
    {
        return Task.FromException<Stream>(e);
    }
}
```
which is a whole lot more (and requiring of more thought) than:
```C#
public override async Task<Stream> GetStreamAsync() => new FileStream(...);
```

I've been seeing this repeatedly in various codebases, where folks get the warning and then think the answer is to use an await, when it's not.

**Comment by CyrusNajmabadi** [comment]
This has often felt like a warning that was more useful early on when people were learning about async/await.   I find it mostly a nuisance now.

**Comment by stephentoub** [comment]
> This has often felt like a warning that was more useful early on when people were learning about async/await. I find it mostly a nuisance now.

Maybe the answer is just downgrade the warning to a suggestion in general?

**Comment by 333fred** [comment]
After discussion, and especially with runtime async, we will be removing this warning entirely.

**Comment by jjonescz** [comment]
> we will be removing this warning entirely.

The warning still seems useful to me (except in interface implementations and method overrides). E.g., when I'm writing tests, need some `await`s there so I make the test `async`, then remove them, then the warning reminds me to make the test synchronous again. That's just the most common example, but similar thing happens in non-test methods as well of course.

Async methods still add overhead (until we turn on runtime async by default) right?

But perhaps it can be an analyzer.

**Comment by aepot** [comment]
That's an unexpected solution. Why not leaving it opt-in?

What can I do to see if I suddenly forgot to remove `async` keyword where it was really not needed? For instance in `async void` method.

Tell me if this method
```c#
async void SomeMethod()
{
    // ...some synchronous code
}
```
is completely the same as this one
```c#
void SomeMethod()
{
    // ...some synchronous code
}
```
after compilation and compiler won't add the async overhead, then I'll be happy.

**Comment by E021ntox** [comment]
This warning is very important, why should it be removed?
It is crucial for keeping the code clean. Please provide an alternative solution.

**Comment by risperdal** [comment]
> That's an unexpected solution. Why not leaving it opt-in?
> 
> What can I do to see if I suddenly forgot to remove `async` keyword where it was really not needed? For instance in `async void` method.
> 
> Tell me if this method
> 
> async void SomeMethod()
> {
>     // ...some synchronous code
> }
> is completely the same as this one
> 
> void SomeMethod()
> {
>     // ...some synchronous code
> }
> after compilation and compiler won't add the async overhead, then I'll be happy.

Any offical clarification for the comment?

**Comment by CyrusNajmabadi** [comment]
> Any offical clarification for the comment?

We moved to this being an IDE warning if you care about this.  But it's not something that the compiler will check.  IDE support was added in: https://github.com/dotnet/roslyn/pull/81835

**Comment by drieseng** [comment]
I know I'm late to the party, but have you considered people using the "latest" version of the SDK to target older runtimes for which there's no runtime async (that will handle "async methods that lack any awaits" more optimally).

Wouldn't it have an option to continue raising CS1998 if the target framework did not feature runtime async support?

**Comment by CyrusNajmabadi** [comment]
@drieseng we now report this through an analyzer. No need for this to be baked into the compiler.

**Comment by hez2010** [comment]
> That's an unexpected solution. Why not leaving it opt-in?
> 
> What can I do to see if I suddenly forgot to remove `async` keyword where it was really not needed? For instance in `async void` method.
> 
> Tell me if this method
> 
> async void SomeMethod()
> {
>     // ...some synchronous code
> }
> is completely the same as this one
> 
> void SomeMethod()
> {
>     // ...some synchronous code
> }
> after compilation and compiler won't add the async overhead, then I'll be happy.

You don' need to remove it. With runtime async your `async void SomeMethod` is as fast as `void SomeMethod` when there's no await in it.

**Comment by aepot** [comment]
> You don' need to remove it. With runtime async your `async void SomeMethod` is as fast as `void SomeMethod` when there's no await in it.

Regarding to code execution speed that's correct but overhead is still present. I suggest to look at some decompiled code for details.

Btw, IDE analyzer is totally fine for me. Thanks to all for the solution.

**Comment by 333fred** [comment]
`async void` in general is an anti-pattern that should be avoided whether possible. 

> With runtime async your `async void` SomeMethod is as fast as `void SomeMethod` when there's no await in it. 

No. The runtime doesn't support `async void`, it will always use a state machine.
